src/v/cloud_storage/segment_meta_cstore.h

```diff
@@ -74,7 +75,7 @@ class segment_meta_materializing_iterator
 
     bool equal(const segment_meta_materializing_iterator& other) const;
 
-    std::unique_ptr<impl> _impl;
+    ss::lw_shared_ptr<impl> _impl;
```

Comment:
@andijcr  do you see any problems make the iterator copyable?

Reply:
not in principle, but I need to recheck the implementation if it could introduce weird artifacts. the commit message cites some std::algorithms. I remember i was able to use std::equal in a test by selecting the implementation that did not copy iterators.
maybe if the failure is only in tests, it's safer to convert the code to raw loops while we work on proper refactor.

Reply:
this seems incorrect, when iterator is copied they will be pointing to the same impl. This way when one copy is incremented the other one will be as well

Reply:
@BenPope does this patch work to fix the stdc++ error? I'm not sure where you ran into this.

```
diff --git a/src/v/cloud_storage/segment_meta_cstore.cc b/src/v/cloud_storage/segment_meta_cstore.cc
index 9490ceaa18..951e11174a 100644
--- a/src/v/cloud_storage/segment_meta_cstore.cc
+++ b/src/v/cloud_storage/segment_meta_cstore.cc
@@ -1120,8 +1120,17 @@ segment_meta_cstore::operator=(segment_meta_cstore&&) noexcept
 bool segment_meta_cstore::operator==(segment_meta_cstore const& oth) const {
     // this overload of std::equal in clang stdlib works with const_iterator
     // because it does not perform a copy of the iterator
-    return size() == oth.size()
-           && std::equal(begin(), end(), oth.begin(), std::equal_to{});
+    if (size() != oth.size()) {
+        return true;
+    }
+    auto this_it = begin();
+    auto oth_it = oth.begin();
+    for (size_t i = 0; i < size(); ++i) {
+        if (*this_it != *oth_it) {
+            return false;
+        }
+    }
+    return true;
 }

 segment_meta_cstore::const_iterator segment_meta_cstore::begin() const {
```

Reply:
The error I got was with libc++ 18.

Something like this might be better :P

Reply:
Cool I updated to that